### PR TITLE
Revamp flash scripts

### PR DESCRIPTION
**Pull request by mbolivar**
This series:

- Fixes minor issues around flashing and DT that I noticed while poking around
- Re-works the flashing scripts to use command line arguments instead of environment variables
- Adds infrastructure to flash based on addresses in DT
- Illustrates DT-based flashing for pyocd and dfu-util (though a bit more is needed to make mcuboot app flashing "easy", which can't be added conveniently for reasons I'll explain if anyone asks)
- Adds some additional polish and convenience (e.g. print all available runners) to zephyr_flash_debug.py

It's meant for when the 1.11 merge window opens up.

@galak, @SebastianBoe, with this, it would be easy to support the features described in https://github.com/zephyrproject-rtos/zephyr/issues/4957. E.g. we could have each boards/common/foo.board.cmake (for foo = openocd, pyocd, dfu-util, etc.) add "foo" to a global property list of available flashers (or debuggers, depending on foo). The flash/CMakeLists.txt file could use that list to generate flash_foo, debug_foo, etc. targets.

However, the right way to flash and debug is under some discussion (e.g. https://github.com/zephyrproject-rtos/zephyr/issues/5153), so I'm not proposing those changes yet, pending decisions there.

### CODE REVIEW

File: boards/arm/v2m_beetle/doc/v2m_beetle.rst
Comment (dbkinder) [inline boards/arm/v2m_beetle/doc/v2m_beetle.rst]: I see you removed "flash" from the goals, but i don't see any replacement text for how to flash?  Perhaps your comment "the right way to flash and debug is under some discussion" means you don't have replacement text yet...

**Review by dbkinder — COMMENTED** [review]
I presume your comment, "the right way to flash and debug is under some discussion" is why there's no replacement text for how to flash in the documentation change you made...

File: boards/arm/v2m_beetle/doc/v2m_beetle.rst
Comment (mbolivar) [reply boards/arm/v2m_beetle/doc/v2m_beetle.rst]: Sorry to have been unclear. I tried to explain this change in the commit log:

```
This board doesn't support the 'flash' goal, which was mistakenly
included during the CMake conversion.
```

This is one of the "minor issues around flashing and DT that I noticed while poking around".

**Comment by mbolivar** [comment]
@andrewboie, @carlescufi, @jhedberg and @MaureenHelm as code owners:

Feedback (especially testing on your platforms!) would be appreciated. @lpereira, you too -- I couldn't get esp32 to build following along with the documentation. Any confirmation that this is working for your boards would be great.

**Comment by mbolivar** [comment]
> though a bit more is needed to make mcuboot app flashing "easy", which can't be added conveniently for reasons I'll explain if anyone asks

Let me add a bit more context here now that I've had some more time.

The basic issue is that mcuboot does not simply chain-load a binary at some flash device offset that's known to both the bootloader and the application linker script, but rather parses and validates a blob which includes digital signatures and other metadata located at that shared flash offset.

So for the zephyr scripts to flash something usable by mcuboot, they need additional command line arguments which can override what binary they flash to specify the output of the mcuboot postprocessor script (imgtool.py) instead of simply assuming zephyr.bin.

This is complicated because not all the flasher scripts use zephyr.bin. For example, the nrfjprog runner uses zephyr.hex.

So while it's a prerequisite to teach the flashers about arbitrary flash addresses, it's not the whole story.

File: cmake/flash/CMakeLists.txt
Comment (carlescufi) [inline cmake/flash/CMakeLists.txt]: Why "kernel-*"? Wouldn't it be better to call it "zephyr-*"? This immediately confused me.

File: scripts/support/runner/core.py
Comment (carlescufi) [inline scripts/support/runner/core.py]: see previous comment

File: cmake/flash/CMakeLists.txt
Comment (SebastianBoe) [inline cmake/flash/CMakeLists.txt]: The below is used elsewhere to solve the same problem.

    $<$<BOOL:${CMAKE_VERBOSE_MAKEFILE}>:--verbose>

I think it makes sense to re-use CMAKE_VERBOSE_MAKEFILE here,
because it is easier for a user if he only needs to learn the name of
one CMake variable to pass --verbose to underlying commands.

See zephyr/arch/x86/CMakeLists.txt

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
I like that configuration has moved from env vars to the cli for the flashers, and that different flashers can extend the CLI with their arguments.

But some of the changes are not obviously positive.

We want each layer to know as little as possible. Previously the flashing layer did not have access to the CONFIG_* values. The CMake layer however already knows what all CONFIG_* values are, so couldn't we avoid coupling the flashing layer to the Kconfig metadata by selectively passing on some CONFIG_* values from the CMake layer to the flashing layer?

I am afraid it could confuse a user if he sees two identical flashing commands that act differently because the flasher parses the .config file and acts differently.

EDIT: This is an example of where the runner is accessing CONFIG_* information directly.
https://github.com/zephyrproject-rtos/zephyr/pull/5232/files#diff-625597f38023505c804e54e3ab18ffcdR359

**Comment by carlescufi** [comment]
First of all thanks @mbolivar for this PR. It addresses and reworks many of our concerns so far, especially the env variables dark magic and instead sets a solid foundation for command-line reusable tools for flashing and running.

@SebastianBoe I actually sort of disagree with your comments. I see no real drawback in accessing the Kconfig options from the flashers, because at the end of the day the configuration chosen by the user with Kconfig and DTS is something that needs to live through the whole process of building, flashing and running in order for the tools to have full knowledge of the metadata inside the .elf and .bin/.hex files. What would selective forwarding of certain Kconfig (and DTS?) options give us instead of having those available all through the process? I imagine for example that there will be new options in the future implemented as Python scripts and that will also require extensive information that can only be obtained from DTS and Kconfig.

**Comment by SebastianBoe** [comment]
It might be possible / desirable to get rid of the CMake variables that are defined by board build scripts, and read by flasher build scripts.

We essentially have 2 interfaces to the flasher scripts, one that is a very thin
layer of CMake code that uses CMake variables. And then the one defined by
argparse. The CMake layer duplicates a lot of information with the argparse layer,
like whether an argument is optional or required.

Getting rid of the CMake layer would make everything more consistent and
easier to maintain.

See this as an example of how this could be done:
https://github.com/SebastianBoe/zephyr/commit/e8e9fb093e830b8d10c2b84389344d9c97b8450d

Now, not all information could be converted to this format, sometimes a board needs to set
a CMake variable so that it can be used in the build multiple places. But that is an exception,
rather than a rule.

@mbolivar : This architecture is inherited from KBuild, so you could dismiss this as out-of-scope if you want.

**Comment by carlescufi** [comment]
> Getting rid of the CMake layer would make everything more consistent and
> easier to maintain.
> 
> See this as an example of how this could be done:
> SebastianBoe/zephyr@e8e9fb0

This reduces the abstraction inside the CMake files by targetting directly the flash Python tool instead of storing state in intermediate, build-system only variables.
+1 to that

File: scripts/support/runner/core.py
Comment (SebastianBoe) [inline scripts/support/runner/core.py]: CMake parses values from ${DOTCONFIG} which is located somewhere else than auto.conf.

I don't know what the difference between auto.conf and ${DOTCONFIG} is. But I'm sure
we want CONFIG_* vars to be the same in the CMake namespace and in this namespace
so we should set them to the same file.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]

File: cmake/flash/CMakeLists.txt
Comment (mbolivar) [reply cmake/flash/CMakeLists.txt]: The variable KERNEL_ELF_NAME is already present in CMake before this series. I'm using the same name in Python for consistency. Do you think we should change it?

File: scripts/support/runner/core.py
Comment (mbolivar) [reply scripts/support/runner/core.py]: Oh, OK, nice catch! That's a good point.

**Comment by lpereira** [comment]
I'm taking a look re: ESP32.

**Comment by mbolivar** [comment]
>> Getting rid of the CMake layer would make everything more consistent and
>> easier to maintain.
>> 
>> See this as an example of how this could be done:
>> SebastianBoe/zephyr@e8e9fb0
> This reduces the abstraction inside the CMake files by targetting directly the flash Python tool instead of storing state in intermediate, build-system only variables.
> +1 to that

I like this quite a bit.

I did this PR with CMake variables to minimize diff in the board files, but I'm happy to teach them about the python script directly (though this might open them up to a bit of churn, hopefully that's OK).

File: scripts/support/runner/core.py
Comment (MaureenHelm) [inline scripts/support/runner/core.py]: The commit message is a bit confusing since your changes don't actually check if a runner implements CONFIG_FLASH_LOAD_OFFSET. Are you suggesting that there will be a new command added for runners that implement it?

**Review by MaureenHelm — COMMENTED** [review]
Thanks @mbolivar , this is really nice. I started thinking recently about how to add flashing capability to the jlink runner and bumped into one of the problems this PR works to solve. I also think it will help with creating a bootable image for mimxrt1050_evk, which has external flashes.

I'm afraid I'm in the middle of some PC issues though, so I won't be able to test on hardware for a few days.

File: scripts/support/runner/core.py
Comment (MaureenHelm) [reply scripts/support/runner/core.py]: Never mind, you answered this in a later commit

File: cmake/flash/CMakeLists.txt
Comment (carlescufi) [reply cmake/flash/CMakeLists.txt]: Yes, actually I think we should. Especially given that in the future we might decide to split the kernel into a different image for security or conformance reasons.

File: cmake/flash/CMakeLists.txt
Comment (mbolivar) [reply cmake/flash/CMakeLists.txt]: Thanks; I wasn't aware of that. I'll fix it.

**Comment by mbolivar** [comment]
Updated.

- rebased (dropping two patches which made v1.10)
- added 9c016ed6b
- fixed s/auto.conf/.config/
- got rid of FLASH_DEBUG_VERBOSE

I need to test this some more tomorrow but want to provide an update for European TZs.

>>> Getting rid of the CMake layer would make everything more consistent and
>>> easier to maintain.
> 
>>> See this as an example of how this could be done:
>>> SebastianBoe/zephyr@e8e9fb0
>> This reduces the abstraction inside the CMake files by targetting directly the flash Python tool instead of storing state in intermediate, build-system only variables.
>> +1 to that
> I like this quite a bit.
> 
> I did this PR with CMake variables to minimize diff in the board files, but I'm happy to teach them about the python script directly (though this might open them up to a bit of churn, hopefully that's OK).

I tried to implement this, and now don't think it's a good idea. Here is why.

We agree that variables which board.cmake or the application CMakeLists.txt *must* set (i.e. mandatory variables with no default values, like NRF_FAMILY as pointed out by @SebastianBoe) can be safely converted to supply script arguments directly.

I think these cases should *not* be converted:

- User-settable variables (set at generation time or maybe "make" time, like board IDs / serial ports). Removing CMake variables means users cannot write "cmake -DVARIABLE=value".  Instead, they must edit CMake files, which is a worse experience, and hinders automation.

- Variables that board.cmake/application CMakeLists.txt may override/customize -- I don't mean set a flag, I mean change one non-boolean value to another. With CMake variables, it's easy for foo.board.cmake to test if FOO_SOME_VAR is unset, then supply default behavior for SOME_VAR. Adding options as strings directly to BOARD_RUNNER_ARGS_foo means having to do really ugly string processing instead, which is a worse implementation.

I analyzed the CMake variables which affect runners assuming these criteria:

https://gist.github.com/mbolivar/eb1d22872a4d6dc0b5022c08effe6d2d

9 can be converted, 27 cannot, and 2 I don't know. Moreover, in all but two cases (dfu-util and nrfjprog), the majority of each runner's variables cannot be converted.

And even cases like NRF_FAMILY and DFUUTIL_ALT are easier for board porters than adding options manually. It's easy for dfu-util.board.cmake to check if DFUUTIL_ALT is missing and supply a useful error message otherwise. That is easy for someone writing a board.cmake to understand. It's harder to understand what went wrong when a script they never heard of complains about a missing --pid option at runtime.

Given that most options can't be converted (so adding special cases makes the interface less consistent), and that there are advantages to keeping variables even in cases when they can, I argue that it's better to leave this as it is.

**Comment by mbolivar** [comment]
> User-settable variables (set at generation time or maybe "make" time, like board IDs / serial ports). Removing CMake variables means users cannot write "cmake -DVARIABLE=value". Instead, they must edit CMake files, which is a worse experience, and hinders automation.

I guess I could support something like `RUNNER_foo_ARGS="--bar=quux --baz"` and append them on the global property, but that seems worse too.

**Comment by SebastianBoe** [comment]
> I guess I could support something like RUNNER_foo_ARGS="--bar=quux --baz" and append them on the global property, but that seems worse too.

I was thinking that we could introduce a convention for targets so users can do

make <target>_args="--some-flag"

So it would be

make flash_args="--serial-nr=58076"

<target>_args are placed last on the command line and the script ensures that later arguments override earlier arguments, just like gcc usually does.

**Comment by SebastianBoe** [comment]
> With CMake variables, it's easy for foo.board.cmake to test if FOO_SOME_VAR is unset, then supply default behavior for SOME_VAR. 

We can still support this use-case.

The app CMakeLists.txt can append flags.
The foo.board.cmake can prepend flags.

Conflicting flags are dealt with by the script e.g.

"-fno-exceptions -fexceptions" is interpreted as "-fexceptions"

The main benefit is perhaps that it is easier for users to discover options, as they are discoverable through the familiar "--help" flag.

They can do;

make flash_args=--help

and then

make flash_args=--openocd-gdb-port=2771

**Comment by SebastianBoe** [comment]
A patchset like this is painful to maintain out of tree, so we can revisit this later if you are still sceptical @mbolivar .

**Comment by mbolivar** [comment]
> We can still support this use-case.
> 
> The app CMakeLists.txt can append flags.
> The foo.board.cmake can prepend flags.
> 
> Conflicting flags are dealt with by the script e.g.
> 
> "-fno-exceptions -fexceptions" is interpreted as "-fexceptions"

I did think about "last one wins", but it doesn't work properly for cases like --dfuse-option, which may be given multiple times with cumulative effect. The fact that a case like that exists now means we still need the special-purpose string handling code in CMake.

> 
> The main benefit is perhaps that it is easier for users to discover options, as they are discoverable through the familiar "--help" flag.
> 
> They can do;
> 
> make flash_args=--help
> 
> and then
> 
> make flash_args=--openocd-gdb-port=2771

How can you do this? I thought we discussed that the only known way with cmake to pass a variable at runtime (in general, that would work with Ninja) was through the environment. Did I miss something? If so, that changes a lot of things for me...

> A patchset like this is painful to maintain out of tree, so we can revisit this later if you are still sceptical @mbolivar .

I'm OK to try to resolve this discussion before the merge; I hope we are getting close.

**Comment by mbolivar** [comment]
> I did think about "last one wins", but it doesn't work properly for cases like --dfuse-option, which may be given multiple times with cumulative effect.

Let me think about getting rid of appendable options and otherwise verify that argparse can work in a "lst one wins" style. I think appending is useful, but being able to use the --help strings is indeed a nice thing.

File: cmake/flash/CMakeLists.txt
Comment (mbolivar) [reply cmake/flash/CMakeLists.txt]: @carlescufi: I looked into this further, and that change would be user-visible, since cmake/app/boilerplate.cmake gets all of these names from CONFIG_KERNEL_BIN_NAME. That would need to become CONFIG_ZEPHYR_BIN_NAME for consistency, which would break out of tree configs.

Given that and since this naming scheme is pre-existing, I'd like to regard this as out of scope for this series if OK with you.

Comment (carlescufi) [reply cmake/flash/CMakeLists.txt]: @mbolivar I understand. I will create a separate issue to deal with this, thanks for looking into it.

**Comment by SebastianBoe** [comment]
> How can you do this? I thought we discussed that the only known way with cmake to pass a variable at runtime (in general, that would work with Ninja) was through the environment. Did I miss something? If so, that changes a lot of things for me...

My mistake, this doesn't work with Ninja.

Perhaps the environment is the cleanest way to do this then. I think it would help usability if a user
could do:

$ flash_args="--help" ninja flash

**Comment by mbolivar** [comment]
A new update is available. I've tested on as many backends as I can (I've flashed arduino_101, 96b_carbon, nrf52_pca10040, frdm_k64f, disco_l475_iot1, and lightly debugged with those boards that support it). I used the (new in this version) DEBUG global in core.py to eyeball the command lines for other targets, and they seem sane.

The main difference from the previous version is that I've deleted almost all the CMake variable usage in boards/common/xxx.board.cmake, as discussed with @carlescufi and @SebastianBoe. The new extension function board_runner_args() should be used to add arguments directly instead in most cases.

You will see a few variables left (especially in openocd.board.cmake) where I thought they were helpful.

I've also added and removed a couple of less important patches.

> Perhaps the environment is the cleanest way to do this then. I think it would help usability if a user
> could do:
> 
> $ flash_args="--help" ninja flash

Sigh, I really want to get away from the environment, but I also hear what you're saying here. I'm hopeful if we keep working on this we'll find a solution that doesn't require the environment. Hopefully this series is a step towards it.

**Comment by mbolivar** [comment]
> A new update is available. I've tested on as many backends as I can (I've flashed arduino_101, 96b_carbon, nrf52_pca10040, frdm_k64f, disco_l475_iot1, and lightly debugged with those boards that support it).

Scratch that -- I botched the rebase. Another version incoming

**Comment by mbolivar** [comment]
> Scratch that -- I botched the rebase. Another version incoming

Nevermind, this should be fine -- I ran a last test in a shell with the wrong ZEPHYR_BASE. Silly environment variables :)

**Review by SebastianBoe — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**
